src/preload.cpp

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright The async-profiler authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <signal.h>
+#include <dlfcn.h>
+#include "asprof.h"
+
+#ifdef __linux__
+
+int (*sigaction_func)(int, const struct sigaction *, struct sigaction *) = NULL;
+
+extern "C" WEAK DLLEXPORT
+int sigaction(int signum, const struct sigaction *act, struct sigaction *oldact) {
+    if (sigaction_func == NULL) {
+        sigaction_func = (decltype(sigaction_func))dlsym(RTLD_NEXT, "sigaction");
+    }
+    return sigaction_func(signum, act, oldact);
+}
+
+#else
+
+extern "C" DLLEXPORT
+int mac_sigaction(int signum, const struct sigaction *act, struct sigaction *oldact) {
+    return sigaction(signum, act, oldact);
+}
+
+__attribute__((used)) static struct {
+    const void *replacement;
+    const void *original;
+} interposers[] __attribute__((section("__DATA,__interpose"))) = {
```

Comment:
Well... it's a hack )
I'd rather prefer to leave current behavior than to maintain such tricks. Can we detect preloaded library somehow differently? Maybe checking order of loaded libraries? I know we discussed that, but I still hope for a cleaner solution. I'm inclined to accept the fact that checking LD_PRELOAD/DYLD_INSERT_LIBRARIES environment variable is a good enough approach.